[BUG] hypothesis.errors.InvalidDefinition: Type APIWorkflow defines no rules

### Checklist

- [x] I checked the [FAQ section](https://schemathesis.readthedocs.io/en/stable/faq.html#frequently-asked-questions) of the documentation
- [x] I looked for similar issues in the [issue tracker](https://github.com/schemathesis/schemathesis/issues)
- [x] I am using the latest version of Schemathesis

### Describe the bug

Following docs for [Lazy schema loading](url) I cannot get the state_machine to run. I simplified it down to following code:

```
import schemathesis
from pytest import fixture


@fixture
def state_machine():
    schema = schemathesis.from_dict({
        "openapi": "3.0.2",
        "info": {},
        "paths": {}
    })
    return schema.as_state_machine()


def test_statefully(state_machine):
    state_machine.run()
```

```
______________________________________________________________________________________________________________ test_statefully _______________________________________________________________________________________________________________

state_machine = <class 'schemathesis.specs.openapi.stateful.APIWorkflow'>

    def test_statefully(state_machine):
>       state_machine.run()

tests/api/test_schemathesis.py:16: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
venv/lib/python3.11/site-packages/schemathesis/stateful/state_machine.py:60: in run
    return run_state_machine_as_test(cls, settings=settings)
venv/lib/python3.11/site-packages/schemathesis/stateful/__init__.py:133: in run_state_machine_as_test
    return _run_state_machine_as_test(state_machine_factory, settings=settings, _min_steps=2)
venv/lib/python3.11/site-packages/hypothesis/stateful.py:217: in run_state_machine_as_test
    run_state_machine(state_machine_factory)
venv/lib/python3.11/site-packages/hypothesis/stateful.py:113: in run_state_machine
    @given(st.data())
venv/lib/python3.11/site-packages/schemathesis/stateful/state_machine.py:41: in __init__
    super().__init__()  # type: ignore
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = APIWorkflow({})

    def __init__(self) -> None:
        if not self.rules():
>           raise InvalidDefinition(f"Type {type(self).__name__} defines no rules")
E           hypothesis.errors.InvalidDefinition: Type APIWorkflow defines no rules
E           Falsifying example:

venv/lib/python3.11/site-packages/hypothesis/stateful.py:248: InvalidDefinition
```

### To Reproduce

🚨 **Mandatory** 🚨: Steps to reproduce the behavior:

1. create a test file with above code
2. `pytest -k test_statefully`
3. See error

Please include a minimal API schema causing this issue:

```yaml
{
        "openapi": "3.0.2",
        "info": {},
        "paths": {}
}
```

### Expected behavior

Able to run state_machine as documented.

### Environment

```
- OS: Fedora 39
- Python version: 3.11 venv
- Schemathesis version: 3.24.3
- Spec version: Open API 3.0.2
```

### Additional context

Originally I used real OpenApi schema which I am able to lazily load using `from_pytest_fixture()` loading using `schemathesis.from_uri()`, but the issue does not seem to be in the schema itself. I tried running with `state_machine.run(settings=settings(max_examples=1))` but that does not help.

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -13,6 +13,7 @@ Changelog
 **Changed**
 
 - **Python**: Explicitly note that combining ``@schema.given`` with explicit examples from the spec is not supported. :issue:`1217`
+- Clarify error message when a state machine has no transitions. :issue:`1992`
 
 **Fixed**
 

--- a/docs/stateful.rst
+++ b/docs/stateful.rst
@@ -7,6 +7,10 @@ It works great when you need to quickly verify that your operations properly val
 
 With stateful testing, Schemathesis combines multiple API calls into a single test scenario and tries to find call sequences that fail.
 
+Stateful tests in Schemathesis rely on Open API links to function, as they are designed to target stateful transitions between API endpoints.
+Unlike stateless tests, which verify individual endpoints in isolation, stateful tests require these links to sequence API calls logically. 
+Ensure your schema includes Open API links to leverage stateful testing effectively.
+
 Why is it useful?
 -----------------
 

--- a/src/schemathesis/constants.py
+++ b/src/schemathesis/constants.py
@@ -32,6 +32,11 @@ SERIALIZERS_SUGGESTION_MESSAGE = (
     "and Schemathesis will be able to make API calls with this media type. \n"
     "See https://schemathesis.readthedocs.io/en/stable/how.html#payload-serialization for more information."
 )
+NO_LINKS_ERROR_MESSAGE = (
+    "Stateful testing requires at least one OpenAPI link in the schema, but no links detected. "
+    "Please add OpenAPI links to enable stateful testing or use stateless tests instead. \n"
+    "See https://schemathesis.readthedocs.io/en/stable/stateful.html#how-to-specify-connections for more information."
+)
 EXTENSIONS_DOCUMENTATION_URL = "https://schemathesis.readthedocs.io/en/stable/extending.html"
 ISSUE_TRACKER_URL = (
     "https://github.com/schemathesis/schemathesis/issues/new?"

--- a/src/schemathesis/stateful/state_machine.py
+++ b/src/schemathesis/stateful/state_machine.py
@@ -5,8 +5,10 @@ from dataclasses import dataclass
 from typing import TYPE_CHECKING, Any, Callable, ClassVar
 
 from hypothesis.stateful import RuleBasedStateMachine
+from hypothesis.errors import InvalidDefinition
 
-from ..constants import NOT_SET
+from ..constants import NOT_SET, NO_LINKS_ERROR_MESSAGE
+from ..exceptions import UsageError
 from ..models import Case, APIOperation, CheckFunction
 
 if TYPE_CHECKING:
@@ -38,7 +40,12 @@ class APIStateMachine(RuleBasedStateMachine):
     schema: BaseSchema
 
     def __init__(self) -> None:
-        super().__init__()  # type: ignore
+        try:
+            super().__init__()  # type: ignore
+        except InvalidDefinition as exc:
+            if "defines no rules" in str(exc):
+                raise UsageError(NO_LINKS_ERROR_MESSAGE) from None
+            raise
         self.setup()
 
     def _pretty_print(self, value: Any) -> str:

--- a/test/specs/openapi/test_stateful.py
+++ b/test/specs/openapi/test_stateful.py
@@ -1,8 +1,10 @@
 import pytest
 from hypothesis import HealthCheck, Phase, settings
+from hypothesis.errors import InvalidDefinition
 
 import schemathesis
-from schemathesis.exceptions import CheckFailed
+from schemathesis.constants import NO_LINKS_ERROR_MESSAGE
+from schemathesis.exceptions import CheckFailed, UsageError
 from schemathesis.specs.openapi.stateful.links import make_response_filter, match_status_code
 from schemathesis.stateful.state_machine import StepResult
 from src.schemathesis.models import CaseSource, Check, Status
@@ -279,3 +281,25 @@ def test_history(testdir, app_schema, base_url, response_factory, method):
     # Then they should store all history
     assert serialized.history[0].case.verbose_name == "PATCH /api/users/{user_id}"
     assert serialized.history[1].case.verbose_name == "POST /api/users/"
+
+
+@pytest.mark.openapi_version("3.0")
+@pytest.mark.operations("success")
+def test_no_transitions_error(app_schema):
+    schema = schemathesis.from_dict(app_schema)
+    state_machine_cls = schema.as_state_machine()
+
+    with pytest.raises(UsageError, match=NO_LINKS_ERROR_MESSAGE):
+        state_machine_cls()
+
+
+@pytest.mark.openapi_version("3.0")
+@pytest.mark.operations("create_user", "get_user", "update_user")
+def test_settings_error(app_schema):
+    schema = schemathesis.from_dict(app_schema)
+
+    class Workflow(schema.as_state_machine()):
+        settings = settings(max_examples=5)
+
+    with pytest.raises(InvalidDefinition):
+        Workflow()

